### PR TITLE
feat: mask sdk (initial pr)

### PR DESCRIPTION
**Pull request by Jack-Works**
mf-4346

Now only the following methods are supported:

```ts
    EthereumMethodType.ETH_GET_CODE,
    EthereumMethodType.ETH_GAS_PRICE,
    EthereumMethodType.ETH_BLOCK_NUMBER,
    EthereumMethodType.ETH_GET_BALANCE,
    EthereumMethodType.ETH_GET_BLOCK_BY_NUMBER,
    EthereumMethodType.ETH_GET_BLOCK_BY_HASH,
    EthereumMethodType.ETH_GET_TRANSACTION_BY_HASH,
    EthereumMethodType.ETH_GET_TRANSACTION_RECEIPT,
    EthereumMethodType.ETH_GET_TRANSACTION_COUNT,
    EthereumMethodType.ETH_GET_FILTER_CHANGES,
    EthereumMethodType.ETH_NEW_PENDING_TRANSACTION_FILTER,
    EthereumMethodType.ETH_ESTIMATE_GAS,
    EthereumMethodType.ETH_CALL,
    EthereumMethodType.ETH_GET_LOGS,
```

Click the button (not the final UI, only for testing) will inject Mask SDK into the current page

I want to merge this PR at the current state, and more features come in future PRs.

### CODE REVIEW

File: packages/mask-sdk/main/index.ts [hunk 1]
```diff
@@ -1,23 +1,27 @@
...
-        if (process.env.NODE_ENV === 'development') {
-            // @ts-expect-error dev only
-            MaskSDK.reload = () => globalThis.dispatchEvent(new Event('mask-sdk-reload'))
-        }
-    } catch {}
-
+    if (init.debuggerMode) {
+        Object.assign(MaskSDK, { reload: () => contentScript.reload() })
+    }
     Reflect.set(globalThis, 'Mask', MaskSDK)
```
Comment (guanbinrui) [inline packages/mask-sdk/main/index.ts:19]: We'd inject ethereum as well.

Comment (guanbinrui) [inline packages/mask-sdk/main/index.ts:19]:
```suggestion
    Reflect.set(globalThis, 'masknet', MaskSDK)
```

File: packages/mask/background/tasks/Cancellable/InjectContentScripts_declarative.ts [hunk 2]
```diff
@@ -25,8 +19,7 @@ async function unregisterExistingScripts() {
 }
 
 function prepareMainWorldScript(matches: string[]): Scripting.RegisteredContentScript[] {
-    if (Flags.has_firefox_xray_vision) return []
-
+    if (globalThis.navigator?.userAgent.includes('Firefox')) return []
```
Comment (guanbinrui) [inline packages/mask/background/tasks/Cancellable/InjectContentScripts_declarative.ts]: Sniffings.is_firefox

File: packages/mask/background/tasks/Cancellable/InjectContentScripts_imperative.ts [hunk 3]
```diff
@@ -1,122 +1,34 @@
...
-
-        // #region Mask SDK
-        if (Flags.mask_SDK_ready) {
-            const code = process.env.NODE_ENV === 'development' ? await fetchUserScript(maskSDK_URL) : await maskSDK
-            browser.tabs.executeScript(arg.tabId, { ...detail, code }).catch(HandleError(arg))
-        }
-        // #endregion
-        injectContentScript(arg.tabId, arg.frameId).catch(HandleError(arg))
+    if (globalThis.navigator?.userAgent.includes('Firefox')) {
+        browser.tabs.executeScript(arg.tabId, { ...detail, file: injectedScriptURL }).catch(err)
```
Comment (guanbinrui) [inline packages/mask/background/tasks/Cancellable/InjectContentScripts_imperative.ts:24]:
```suggestion
        await browser.tabs.executeScript(arg.tabId, { ...detail, file: injectedScriptURL }).catch(err)
```

File: packages/mask/background/tasks/Cancellable/InjectContentScripts_imperative.ts [hunk 4]
```diff
@@ -1,122 +1,34 @@
...
+    } else {
+        browser.tabs
```
Comment (guanbinrui) [inline packages/mask/background/tasks/Cancellable/InjectContentScripts_imperative.ts]:
```suggestion
        await browser.tabs
```

File: packages/mask/src/content-script.ts [hunk 5]
```diff
@@ -5,6 +5,14 @@ const loaded = Symbol.for('mask_init_content_script')
...
-    await activateSiteAdaptorUI()
+    const state = await activateSiteAdaptorUI()
+    if (state === 'notFound' || state === 'needMaskSDK') {
+        // Not found means this is not accepted by any site adaptor.
+        // This can happens in the following cases:
+        // - User clicked the connect button in the popup, thus we inject the content script.
+        // - We have permission for this site (granted previously?) so the user expect Mask to work here.
+        const { startMaskSDK } = await import('./extension/mask-sdk/index.js')
+        startMaskSDK()
+    }
```
Comment (guanbinrui) [inline packages/mask/src/content-script.ts:16]: await import(/* webpackMode: 'eager' */ './setup.mask-sdk.js')

File: packages/mask/src/extension/mask-sdk/index.ts [hunk 6]
```diff
@@ -5,10 +5,15 @@ import Services from '../service.js'
...
-    let meta: unknown = undefined
-    if (currentSiteContext) meta = (await Services.ThirdPartyPlugin.getHostedMeta(currentSiteContext))?.[1]
-    // TODO: listen to SiteContext connected status
-    return maskSDK.request_init({ context: { connected: !!currentSiteContext, meta } })
+export async function startMaskSDK() {
+    let meta
+    if (currentSiteContext) {
+        const data = await Services.ThirdPartyPlugin.getHostedMeta(currentSiteContext)
+        if (data) meta = new Map([data])
+    }
```
Comment (guanbinrui) [inline packages/mask/src/extension/mask-sdk/index.ts]:
```ts
const data = currentSiteContext ? await Services.ThirdPartyPlugin.getHostedMeta(currentSiteContext) : undefined
const meta = data ? new Map([ data ]) : undefined
```

File: packages/mask/src/extension/mask-sdk/bridge/index.ts [hunk 7]
```diff
@@ -1,11 +1,83 @@
...
+    // https://docs.metamask.io/wallet/reference/eth_subscribe/
+}
+async function eth_request(request: unknown) {
+    if (typeof request !== 'object' || request === null) throw invalidRequest()
+
+    const method = (request as any).method
+    if (typeof method !== 'string' || !method) throw invalidMethod()
+    if (!(method in impl)) throw unknownMethod(method)
+
+    let params: unknown
```
Comment (guanbinrui) [inline packages/mask/src/extension/mask-sdk/bridge/index.ts]: const { params } = request

File: packages/mask/src/extension/mask-sdk/bridge/index.ts [hunk 8]
```diff
@@ -1,11 +1,83 @@
...
+    if (params && params !== undefined) {
+        if (typeof params !== 'object' || !params) throw invalidParams()
```
Comment (guanbinrui) [inline packages/mask/src/extension/mask-sdk/bridge/index.ts]: Array.isArray(params)

File: packages/mask/src/extension/mask-sdk/bridge/index.ts [hunk 9]
```diff
@@ -1,11 +1,83 @@
...
+    }
+
+    return (impl as any)[method](params)
+}
+function invalidRequest() {
+    throw new MaskEthereumProviderRpcError({
+        code: -32600,
```
Comment (guanbinrui) [inline packages/mask/src/extension/mask-sdk/bridge/index.ts:47]: Give an enum and use the more readable name.

File: packages/mask-sdk/shared/index.ts [hunk 10]
```diff
@@ -1,6 +1,9 @@
+/// <reference path="../dist/public-api.d.ts" />
 export interface BridgeAPI {
     persona_sign_web3(message: string): Promise<string>
```
Comment (guanbinrui) [inline packages/mask-sdk/shared/index.ts]:
```suggestion
    persona_signMessage(message: string): Promise<string>
```

File: packages/mask/background/utils/injectScript.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
+import { memoize } from 'lodash-es'
+
+export const injectedScriptURL = '/js/injected-script.js'
+export const maskSDK_URL = '/js/mask-sdk.js'
+const contentScriptURL = '/generated__content__script.html'
+
+export async function evaluateContentScript(tabId: number | undefined, frameId?: number) {
```
Comment (guanbinrui) [inline packages/mask/background/utils/injectScript.ts:7]: Can we load any plugin at the runtime by using this code?

Comment (Jack-Works) [reply packages/mask/background/utils/injectScript.ts:7]: I strongly object to this idea although this is technically possible in Manifest V2 when we don't have our codebase secured.

File: packages/mask-sdk/main/index.ts (hunk 1 above)
Comment (Jack-Works) [reply packages/mask-sdk/main/index.ts:19]: this is our public API entry point. Why do you want it to name "masknet"? it does not look good.

File: packages/mask/src/content-script.ts (hunk 5 above)
Comment (Jack-Works) [reply packages/mask/src/content-script.ts:16]: I don't want to add an eager load for this one, it is not frequently used currently. The Site Adaptor is different. We may add it in the future.

File: packages/mask/src/extension/mask-sdk/bridge/index.ts [hunk 12]
```diff
@@ -1,11 +1,81 @@
...
 
 export const maskSDKServer: BridgeAPI = {
-    async persona_sign_web3(message) {
+    async persona_signMessage(message) {
         return Services.Identity.signWithPersona(SignType.Message, String(message))
     },
+    eth_request,
+    async reload() {
+        if (process.env.NODE_ENV === 'production') return
+        Services.SiteAdaptor.attachMaskSDKToCurrentActivePage
```
Comment (guanbinrui) [inline packages/mask/src/extension/mask-sdk/bridge/index.ts]:
```suggestion
        await Services.SiteAdaptor.attachMaskSDKToCurrentActivePage()
```

**Pull request merged**